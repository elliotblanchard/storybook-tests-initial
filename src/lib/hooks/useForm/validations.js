const minMessage = rule =>
  `At least ${rule} character${rule > 1 ? 's' : ''} ${rule > 1 ? 'are' : 'is'} required.`;
const maxMessage = rule =>
  `At most ${rule} character${rule > 1 ? 's' : ''} ${rule > 1 ? 'are' : 'is'} allowed.`;
const emailMessage = () => 'A valid email is required.';
const upperMessage = () => 'Only strings can be made uppercase.';
const lowerMessage = () => 'Only strings can be made lowercase.';

export const min = (rule, message) => [
  value => [value, value.length >= rule],
  message || minMessage(rule),
];
export const max = (rule, message) => [
  value => [value, value.length <= rule],
  message || maxMessage(rule),
];
export const email = message => [
  value => [
    value,
    /^[a-zA-Z0-9.!#$%&â€™*+/"=?^_`{|}~-]+@[a-zA-Z0-9-]+(?:\.[a-zA-Z0-9-]+)*$/.test(value),
  ],
  message || emailMessage(),
];

const caseChange = (value, transformType) => {
  const isString = typeof value === 'string';
  return [isString ? value[transformType]() : value, isString];
};
export const upper = message => [
  value => caseChange(value, 'toUpperCase'),
  message || upperMessage,
];
export const lower = message => [
  value => caseChange(value, 'toLowerCase'),
  message || lowerMessage,
];
