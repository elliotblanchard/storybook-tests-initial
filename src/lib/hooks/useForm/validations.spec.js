import { min, max, email, lower, upper } from './validations';

describe('validations', () => {
  it('will validate for minimum length', () => {
    expect(min(5)[0]('string longer than 5')).toEqual(['string longer than 5', true]);
    expect(min(5)[0]('four')).toEqual(['four', false]);
    expect(min(2)[0]([1, 2, 3])).toEqual([[1, 2, 3], true]);
    expect(min(2)[0]([1])).toEqual([[1], false]);
  });

  it('will provide a message for minimum length', () => {
    expect(min(5)[1]).toMatchSnapshot();
    expect(min(1)[1]).toMatchSnapshot();
  });

  it('will validate for maximum length', () => {
    expect(max(5)[0]('string longer than 5')).toEqual(['string longer than 5', false]);
    expect(max(5)[0]('four')).toEqual(['four', true]);
    expect(max(2)[0]([1, 2, 3])).toEqual([[1, 2, 3], false]);
    expect(max(2)[0]([1])).toEqual([[1], true]);
  });

  it('will provide a message for maximum length', () => {
    expect(max(6)[1]).toMatchSnapshot();
    expect(max(1)[1]).toMatchSnapshot();
  });

  it('will format strings to be uppercase', () => {
    expect(upper()[0]('lower case string')).toEqual(['LOWER CASE STRING', true]);
    expect(upper()[0]('loWer cAse String')).toEqual(['LOWER CASE STRING', true]);
    expect(upper()[0](4)).toEqual([4, false]);
    expect(upper()[0]({})).toEqual([{}, false]);
  });

  it('will show a message when a non-string is passed to upper', () => {
    expect(upper()[1](4)).toMatchSnapshot();
    expect(upper()[1]({})).toMatchSnapshot();
  });

  it('will format strings to be lowercase', () => {
    expect(lower()[0]('UPPER CASE STRING')).toEqual(['upper case string', true]);
    expect(lower()[0]('uPPEr CaSE STRing')).toEqual(['upper case string', true]);
    expect(lower()[0](4)).toEqual([4, false]);
    expect(lower()[0]({})).toEqual([{}, false]);
  });

  it('will show a message when a non-string is passed to lower', () => {
    expect(lower()[1](4)).toMatchSnapshot();
    expect(lower()[1]({})).toMatchSnapshot();
  });

  it('will validate emails', () => {
    const validEmails = [
      'email@example.com',
      'firstname.lastname@example.com',
      'email@subdomain.example.com',
      'firstname+lastname@example.com',
      'email@123.123.123.123',
      '"email"@example.com',
      '1234567890@example.com',
      'email@example-one.com',
      '_______@example.com',
      'email@example.name',
      'email@example.museum',
      'email@example.co.jp',
      'firstname-lastname@example.com',
    ];
    validEmails.forEach(testEmail => expect(email()[0](testEmail)).toEqual([testEmail, true]));
  });

  it('will reject invalid emails', () => {
    const invalidEmails = [
      'plainaddress',
      '#@%^%#$@#$@#.com',
      '@example.com',
      'Joe Smith <email@example.com>',
      'email.example.com',
      'email@example@example.com',
      'email@example.com (Joe Smith)',
    ];
    invalidEmails.forEach(testEmail => expect(email()[0](testEmail)).toEqual([testEmail, false]));
  });
});
