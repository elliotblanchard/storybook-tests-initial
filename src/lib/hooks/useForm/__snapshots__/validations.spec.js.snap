// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`validations will provide a message for maximum length 1`] = `"At most 6 characters are allowed."`;

exports[`validations will provide a message for maximum length 2`] = `"At most 1 character is allowed."`;

exports[`validations will provide a message for minimum length 1`] = `"At least 5 characters are required."`;

exports[`validations will provide a message for minimum length 2`] = `"At least 1 character is required."`;

exports[`validations will show a message when a non-string is passed to lower 1`] = `"Only strings can be made lowercase."`;

exports[`validations will show a message when a non-string is passed to lower 2`] = `"Only strings can be made lowercase."`;

exports[`validations will show a message when a non-string is passed to upper 1`] = `"Only strings can be made uppercase."`;

exports[`validations will show a message when a non-string is passed to upper 2`] = `"Only strings can be made uppercase."`;
